src/odm_sharing/private/rules.py

```diff
@@ -0,0 +1,232 @@
+import sys
+
+from dataclasses import dataclass
+from enum import Enum, EnumMeta
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Union
+
+import pandas as pd
+
+RuleId = int
+
+
+class RuleMode(str, Enum):
+    SELECT = 'select'
+    FILTER = 'filter'
+    GROUP = 'group'
+    SHARE = 'share'
+
+    def __repr__(self) -> str:
+        return self.value
+
+
+class ParseError(Exception):
+    '''schema parsing error'''
+    pass
+
+
+class SchemaCtx:
+    '''Keeps track of the current state of the parsing process. This object
+    should be created at the beginning of the parsing process and its fields
+    updated throughout.'''
+    filename: str
+    row: int  # current row being parsed
+    column: str  # current field being parsed
+
+    def __init__(self, filename: str) -> None:
+        self.filename = filename
+        self.row = 0
+        self.column = ''
+
+    @property
+    def line(self) -> int:
+        '''line number of current row being parsed'''
+        return self.row + 1
+
+
+@dataclass(init=False)
+class Rule:
+    '''A rule mapped from a sharing schema row'''
+    # XXX: should be immutable, but then we can't initialize it in a clean and
+    # dynamic way
+    id: int  # aka ruleID
+    table: str
+    mode: RuleMode
+    key: str
+    operator: str
+    value: str
+
+
+HEADERS = [
+    'ruleID',
+    'table',
+    'mode',
+    'key',
+    'operator',
+    'value',
+    'notes',
+]
+
+FILTER_OPERATORS = set([
+    '<',
+    '<=',
+    '=',
+    '>',
+    '>=',
+    'in',
+])
+
+ALL_MODES = set(RuleMode)
+GROUP_OPERATORS = set(['AND', 'OR'])
+RULE_FIELD_TYPES = Rule.__annotations__
+HEADER_LIST_STR = ','.join(HEADERS)
+
+
+def gen_error(ctx: SchemaCtx, desc: str) -> ParseError:
+    '''returns a ParseError'''
+    col = f',{ctx.column}' if ctx.column else ''
+    msg = f'{ctx.filename}({ctx.line}{col}): {desc}'
+    print('Error: ' + msg, file=sys.stderr)
+    return ParseError(msg)
+
+
+def fail(ctx: SchemaCtx, desc: str) -> None:
+    '''raises a ParseError'''
+    raise gen_error(ctx, desc)
+
+
+def quote(x: str) -> str:
+    return f"'{x}'"
+
+
+def fmt_set(values: Iterable) -> str:
+    '''returns a comma-separated string of the items in `values`'''
+    items = ','.join(map(quote, values))
+    return f'{{{items}}}'
+
+
+def coerce_value(  # type: ignore
+    ctx: SchemaCtx,
+    type_class,
```

Comment:
This should document the type of `type_class` since it does not have a type.

Reply:
done